### DISCUSSION TITLE
Build vs dev performance

### DISCUSSION THREAD

**Discussion by jech33** [xyflow/xyflow#2121] [Q&A]
HI! I noticed that there are performance issues when working in development with react-flow, mostly on nodes that have 3 or more edges connected. BUT, in production, there have not been any performance issues so far. 

What I mean is: 
dev = laggy, freezes. 
production build = smooth, pleasant to use

Has this happened to anyone? If so, how can I fix it? or am I doing something wrong?

Thanks!

**Comment by moklick** [comment]
Hey! Thanks for your feedback. What setup do you have? How many nodes and edges are you rendering? Is it also laggy when dev tools are closed while developing?

**Reply by jech33** [reply]
Hey mocklick! yes! when dev tools are closed is laggy but manageable, just not as smooth as the production build. I'm rendering around 30 nodes and 27 links for the tests I make, but the app sometimes has more or fewer nodes and links, it depends on many factors. 

I'm also using logic to calculate which handles are not connected, I have memoized all custom nodes and links, and also I've verified every function is using `useCallback`

![image](https://user-images.githubusercontent.com/67987594/166565350-dfc4d9b2-7b6d-4e09-ade9-0236b201c124.png)

**Reply by moklick** [reply]
Your app looks pretty nice! Which version are you using? Do you have a repo I could check?

**Reply by jech33** [reply]
Thanks!! I'm using the latest version, because of security reasons of the company I'm working with I cannot share the repo :(. Is there another way you can check?

**Comment by jech33** [comment]
For anyone wondering about this, I think it got fixed with some of the changes in 10.3.0. release. I also removed some unnecessary dependencies on some `useCallbacks` and try to pass any additional functions or info related to nodes and edges in their `data` property. It still has some freeze behavior in the dev environment (you have to play a lot with the nodes to reproduce it 🤣)  but it împroved like a 100%.

I will track here any additional behavior I run into.

**Comment by kuldeepkeshwar** [comment]
@jech33 @moklick  In my case, when dev tool is opened, it almost freezes and manageable without dev tool.   
I have reached this stage after wrapping all callbacks in `useCallback` & nodeTypes in `React.memo`

https://user-images.githubusercontent.com/10448534/181880128-1a0cd9e8-87ec-4c61-9540-5f3758c07f90.mov

> Note : this is dev environment & using v10.3.12

**Reply by jech33** [reply]
Have you tried it after building it in production?

**Reply by kuldeepkeshwar** [reply]
Not yet but I tried to use it multiple browsers, seems like it has something to do with chrome family, worked quite smooth in Firefox
